### PR TITLE
Constructor bounds

### PR DESCRIPTION
**Pull request by xwvvvvwx**
Fixes https://github.com/ethereum/act/issues/66.

Adds preconditions to all constructors ensuring that:

1. All local storage vars are initially zero
1. All environment variables are within the range of their type
1. All external storage is within the range of it's type
1. All calldata variables are within the range of their type

### CODE REVIEW

**Comment by xwvvvvwx** [comment]
[`6bdc0dd`](https://github.com/ethereum/act/pull/84/commits/622694fb8d13b67c1e4b635d7d6c0f79cbae1577) adds a prestate to the constructor definitions in the Coq backend. While at first glance this might seem a little weird (there isn't a local prestate when the constructor is running), this simplifies the code in the SMT backend, and more importantly prepares us for handling constructors that reference or write to external storage in the Coq backend (where a prestate does exist in the external contracts). 

This PR already constrains integer values to zero in the constructor precondiditons, and I will extend that to force bytestrings to be empty and booleans to be false.

cc @endorphin @MrChico does this make sense?

**Comment by xwvvvvwx** [comment]
Added support for polymorphic `Eq` in the AST. This allows comparison of `Bool` and `Bytes` types in act specs. This closes #60.

File: src/K.hs [hunk 1]
```diff
@@ -157,16 +162,21 @@ kExprBool :: Exp Bool -> String
...
 kExprBool (LitBool a) = show a
 kExprBool (BoolVar a) = kVar a
+kExprBool (NEq a b) = "notBool (" <> kExprBool (Eq a b) <> ")"
+kExprBool (Eq (a :: Exp t) (b :: Exp t)) = case eqT @t @Integer of
+  Just Refl -> "(" <> kExprInt a <> " ==Int " <> kExprInt b <> ")"
+  Nothing -> case eqT @t @Bool of
+    Just Refl -> "(" <> kExprBool a <> " ==Bool " <> kExprBool b <> ")"
+    Nothing -> case eqT @t @ByteString of
+      Just Refl -> "(" <> kExprBytes a <> " ==K " <> kExprBytes b <> ")" -- TODO: Is ==K correct?
+      Nothing -> error "Internal Error: invalid expression type"
```
Comment (xwvvvvwx) [inline src/K.hs:179]: @MrChico this is a little clunky, is there some nice trick with the Maybe monad I can use to tidy this up a bit?

Comment (MrChico) [reply src/K.hs:179]: hmm, the fact that you are proceeding in the `Nothing` case makes it a bit tricky... I don't immediately see a cleaner way

**Comment by xwvvvvwx** [comment]
Made some small tweaks to the `K` backend:

- rm the logic that added the invariant to the postconditions (this was incomplete and broken as implemented)
- add indentation in the generated K output
- add support for a few more kinds of storage location

**Comment by xwvvvvwx** [comment]
Reverted the Coq changes after looking over them together with @MrChico. I will remove the local constructor prestate from the SMT backend in a seperate PR to avoid having this one get any more messy than it already is.

Summary of the changes made:

- Adds new preconditions to constructor specs:
  - All environment variables are within the range of their type
  - All external storage is within the range of it's type
  - All calldata variables are within the range of their type
  
- Adds support for polymorphic Eq expressions

- Tweaks to the K backend:
  - rm the logic that added the invariant to the postconditions (this was incomplete and broken as implemented)
  - add indentation in the generated K output
  - add support for a few more kinds of storage location

**Review by MrChico — APPROVED** [review]

**Comment by MrChico** [comment]
Sweet! Looks good!

**Comment by endorphin** [comment]
just got to this. looks good to me, and I guess this doesn't need more review since it's already merged and the coq changes were reverted.

**Pull request merged**